rfd/0032-role-tester.md

```diff
@@ -0,0 +1,94 @@
+
+## Scope
+This document will only be focusing on determining access to SSH nodes. The proposed role tester can be easily extended to determine access to database, application, kubernetes etc. By providing separate queries for different access types, we can also answer more specific queries (e.g. which databases does this user have access to? Which roles allow access to database production as user postgres?).
+
+## Architecture
+```
+                |‾‾‾‾‾‾‾‾‾‾‾‾‾|        |‾‾‾‾‾‾‾‾‾‾‾‾‾‾|
+teleport <----> | role tester | <----> | deductive db |
+                |_____________|        |______________|
+```
+
+For simple use cases and better UX, a provided tctl command can be used to query for access-related questions. For advanced use-cases, a provided admin tool lets users execute Datalog queries directly using a command or in a REPL-like shell.
+
+The role tester layer will use a Rust/Go library (TBD) to run queries on the deductive Datalog database. The results are then returned via CLI to the user.
+
+## Details
+
+### Context
+Datalog is a logic programming language that is a subset of Prolog. A common use case for Datalog is as a query language for deductive databases, which is exactly what we need in order to answer access-related questions. Datalog programs are written as a collection of logical constraints which the compiler uses to compute the solution (using whatever method/algorithm it pleases). These logical constraints are sentences in logical form and are a finite set of **facts**, which are assertions of the world in which we operate; and **rules**, which are sentences that allow us to infer new facts from existing ones. This document will describe a model for representing Teleport's RBAC system as a logical model which we can then query to answer our desired questions.
+
+### Facts
+Facts (or predicates) represent assertions of the world in which we operate.For translating Teleport's RBAC system, we can take everything we know from the role configuration and define what we need.
+
+We will need to determine the labels on the node itself, the role's defined allow/deny labels, the role's define logins, and the user to role pairs.
+
+Fact | Example | Interpretation
+--- | --- | ---
+HasRole(user, role) | HasRole(jean, dev) | User 'jean' has role 'dev'
+NodeHasSSHLabel(node, label_key, label_value) | NodeHasSSHLabel(node-1, environment, staging) | SSH node 'node-1' has the label 'environment:staging'
+RoleAllowedSSHLabel(role, label_key, label_value) | RoleAllowedSSHLabel(dev, environment, staging) | Role 'dev' is allowed access to SSH nodes with label 'environment:staging'
+RoleDeniedSSHLabel(role, label_key, label_value) | RoleDeniedSSHLabel(bad, environment, production) | Role 'bad' is denied access to SSH nodes with label 'environment:production'
+RoleHasSSHLogin(role, login) | RoleHasSSHLogin(admin, root) | Role 'admin' can login as os user 'root' to SSH nodes
```

Comment:
A role can deny SSH logins as well so there probably should be a `RoleAllowsLogin` and `RoleDeniesLogin` similar to labels.

Reply:
Ah ok, I didn't know about this. This is interesting, I might need to adjust a few more rules a well.